File: src/cli/commands/list.ts
/**
 * `list` command (R30, R40). Read-only: enumerates `.claude-profiles/`,
 * loads each manifest, marks the active profile from `.state.json`, and
 * prints either a human table or a structured JSON payload.
 *
 * No lock acquired (R43). May briefly disagree with a concurrent `use`/`new`.
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";
import process from "node:process";

import {
  buildPaths,
  listProfiles,
  type ProfileManifest,
} from "../../resolver/index.js";
import { buildStatePaths, readStateFile } from "../../state/index.js";
import {
  formatStateWarning,
  meaningfulStateWarning,
  relativeTime,
  renderTable,
} from "../format.js";
import { createStyle, resolveNoColor, type OutputChannel } from "../output.js";

export interface ListEntryPayload {
  name: string;
  active: boolean;
  description: string | null;
  extends: string | null;
  includes: string[];
  tags: string[];
  /** Last-materialized ISO timestamp — only set on the active profile. */
  lastMaterialized: string | null;
}

export interface ListPayload {
  profiles: ListEntryPayload[];
  /** Surfaced when state file degraded (Missing is silenced — fresh project). */
  stateWarning: { code: string; detail: string } | null;
}

export interface ListOptions {
  cwd: string;
  output: OutputChannel;
  /** When true, force colour off (additive with NO_COLOR env). Default false. */
  noColor?: boolean;
}

export async function runList(opts: ListOptions): Promise<number> {
  const resolverPaths = buildPaths(opts.cwd);
  const statePaths = buildStatePaths(opts.cwd);
  const names = await listProfiles({ projectRoot: opts.cwd });
  const { state, warning } = await readStateFile(statePaths);

  const entries: ListEntryPayload[] = [];
  for (const name of names) {
    const manifest = await tryLoadManifestSummary(resolverPaths.profilesDir, name);
    const isActive = state.activeProfile === name;
    entries.push({
      name,
      active: isActive,
      description: manifest?.description ?? null,
      extends: manifest?.extends ?? null,
      includes: manifest?.includes ?? [],
      tags: manifest?.tags ?? [],
      lastMaterialized: isActive ? state.materializedAt : null,
    });
  }

  const stateWarning = meaningfulStateWarning(warning);

  if (opts.output.jsonMode) {
    const payload: ListPayload = { profiles: entries, stateWarning };
    opts.output.json(payload);
    return 0;
  }

  if (entries.length === 0) {
    opts.output.print("(no profiles — run `c3p new <name>` to create one)");
  } else {
    const style = createStyle({
      isTty: opts.output.isTty,
      platform: process.platform,
      noColor: resolveNoColor(opts.noColor === true),
    });
    // Decide which optional columns to render. A column is visible if any
    // row has content for it — keeps the layout tight in the common case
    // (no descriptions, no tags) while expanding gracefully when used.
    const showDescription = entries.some((e) => e.description !== null && e.description !== "");
    const showTags = entries.some((e) => e.tags.length > 0);

    const rows: string[][] = entries.map((e) => {
      // Active marker: glyph + bold styling per spec — both for redundancy
      // (some terminals strip ANSI; the `*` survives).
      const marker = e.active ? "*" : " ";
      const nameCell = e.active ? `${marker} ${style.bold(e.name)}` : `${marker} ${e.name}`;
      const cells: string[] = [nameCell];
      if (showDescription) cells.push(e.description ?? "");
      if (showTags) cells.push(e.tags.length > 0 ? `[${e.tags.join(", ")}]` : "");
      // Trailing meta column: extends/includes/last-materialized. Joined
      // with a single space so each piece reads as one continuous flag.
      const metaParts: string[] = [];
      if (e.extends) metaParts.push(`extends=${e.extends}`);
      if (e.includes.length > 0) metaParts.push(`includes=[${e.includes.join(",")}]`);
      if (e.lastMaterialized !== null) {
        metaParts.push(`(materialized ${relativeTime(e.lastMaterialized)})`);
      }
      cells.push(metaParts.join(" "));
      return cells;
    });
    opts.output.print(renderTable(rows));
  }

  if (warning && warning.code !== "Missing") {
    opts.output.warn(formatStateWarning(warning));
  }
  return 0;
}

/**
 * Best-effort manifest read for the list/status display. We don't want a
 * malformed profile.json to blow up `list` — degrade to "no manifest fields
 * shown" rather than aborting. The full `validate` command will surface the
 * details.
 */
async function tryLoadManifestSummary(
  profilesDir: string,
  name: string,
): Promise<ProfileManifest | null> {
  try {
    const raw = await fs.readFile(path.join(profilesDir, name, "profile.json"), "utf8");
    const parsed = JSON.parse(raw);
    if (typeof parsed !== "object" || parsed === null) return null;
    return parsed as ProfileManifest;
  } catch {
    return null;
  }
}

File: go/internal/cli/commands/list.go
package commands

import (
	"fmt"

	"github.com/htxryan/c3p/internal/resolver"
	"github.com/htxryan/c3p/internal/state"
)

// listEntryPayload is the per-row JSON shape emitted under --json.
type listEntryPayload struct {
	Name             string   `json:"name"`
	Active           bool     `json:"active"`
	Description      string   `json:"description"`
	Extends          string   `json:"extends"`
	Includes         []string `json:"includes"`
	Tags             []string `json:"tags"`
	LastMaterialized string   `json:"lastMaterialized"`
}

type listPayload struct {
	Profiles     []listEntryPayload `json:"profiles"`
	StateWarning string             `json:"stateWarning"`
}

// RunList implements the `list` verb. Read-only (R43, no lock).
// Mirrors src/cli/commands/list.ts.
func RunList(opts ListOptions) (int, error) {
	paths := state.BuildStatePaths(opts.Cwd)
	st, err := state.ReadStateFile(paths)
	if err != nil {
		return 2, err
	}
	activeName := ""
	if st.State.ActiveProfile != nil {
		activeName = *st.State.ActiveProfile
	}
	lastMaterialized := ""
	if st.State.MaterializedAt != nil {
		lastMaterialized = *st.State.MaterializedAt
	}

	names, err := resolver.ListProfiles(resolver.DiscoverOptions{ProjectRoot: opts.Cwd})
	if err != nil {
		return 2, err
	}

	entries := make([]listEntryPayload, 0, len(names))
	for _, name := range names {
		mlr, _ := resolver.LoadManifest(profileDir(opts.Cwd, name), name)
		manifest := mlr.Manifest
		entry := listEntryPayload{
			Name:        name,
			Active:      name == activeName,
			Description: manifest.Description,
			Extends:     manifest.Extends,
			Includes:    manifest.Includes,
			Tags:        manifest.Tags,
		}
		if entry.Includes == nil {
			entry.Includes = []string{}
		}
		if entry.Tags == nil {
			entry.Tags = []string{}
		}
		if entry.Active {
			entry.LastMaterialized = lastMaterialized
		}
		entries = append(entries, entry)
	}

	stateWarning := ""
	if st.Warning != nil {
		stateWarning = string(st.Warning.Code)
		if st.Warning.Detail != "" {
			stateWarning = stateWarning + ": " + st.Warning.Detail
		}
	}

	if opts.Output.JSONMode() {
		opts.Output.JSON(listPayload{Profiles: entries, StateWarning: stateWarning})
		return 0, nil
	}

	if len(entries) == 0 {
		opts.Output.Print("(no profiles found in .claude-profiles/)")
		return 0, nil
	}

	hasDescription, hasTags := false, false
	for _, e := range entries {
		if e.Description != "" {
			hasDescription = true
		}
		if len(e.Tags) > 0 {
			hasTags = true
		}
	}

	rows := [][]string{}
	header := []string{"NAME"}
	if hasDescription {
		header = append(header, "DESCRIPTION")
	}
	if hasTags {
		header = append(header, "TAGS")
	}
	header = append(header, "META")
	rows = append(rows, header)
	for _, e := range entries {
		nameCell := e.Name
		if e.Active {
			nameCell = "* " + e.Name
		} else {
			nameCell = "  " + e.Name
		}
		row := []string{nameCell}
		if hasDescription {
			row = append(row, e.Description)
		}
		if hasTags {
			row = append(row, joinTags(e.Tags))
		}
		row = append(row, formatListMeta(e))
		rows = append(rows, row)
	}
	// Use the package-internal renderTable from cli — but commands can't
	// import cli. We render inline here.
	opts.Output.Print(renderTable(rows))
	if stateWarning != "" {
		opts.Output.Warn(fmt.Sprintf("note: state file degraded — %s", stateWarning))
	}
	return 0, nil
}

func formatListMeta(e listEntryPayload) string {
	parts := []string{}
	if e.Extends != "" {
		parts = append(parts, "extends "+e.Extends)
	}
	if len(e.Includes) > 0 {
		parts = append(parts, fmt.Sprintf("includes %d", len(e.Includes)))
	}
	if e.Active && e.LastMaterialized != "" {
		parts = append(parts, "last "+e.LastMaterialized)
	}
	if len(parts) == 0 {
		return ""
	}
	out := parts[0]
	for _, p := range parts[1:] {
		out += " · " + p
	}
	return out
}

func joinTags(tags []string) string {
	out := ""
	for i, t := range tags {
		if i > 0 {
			out += ","
		}
		out += t
	}
	return out
}
